be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala

```diff
@@ -72,8 +72,59 @@ sealed class RollCallHandler(dbRef: => AskableActorRef) extends MessageHandler {
   }
 
   def handleCloseRollCall(rpcMessage: JsonRpcRequest): GraphMessage = {
+    // We first create the three (independent) futures
     val ask: Future[GraphMessage] = dbAskWritePropagate(rpcMessage)
-    Await.result(ask, duration) match {
+
+    // Creates a channel for each attendee (of name /root/lao_id/social/PublicKeyAttendee), returns a Future[GraphMessage]
+    def createAttendeeChannels(data: CloseRollCall, rpcMessage: JsonRpcRequest): Future[GraphMessage] = {
+      val listAttendeeChannels: List[(Channel, ObjectType.ObjectType)] = data.attendees.map(attendee => (generateSocialChannel(rpcMessage.getParamsChannel, attendee), ObjectType.CHIRP))
+      val askCreateChannels: Future[GraphMessage] = (dbActor ? DbActor.CreateChannelsFromList(listAttendeeChannels)).map{
+        case DbActorAck() => Left(rpcMessage)
+        case DbActorNAck(code, description) => Right(PipelineError(code, description, rpcMessage.id))
+        case _ => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswer, rpcMessage.id))
+      }
+      askCreateChannels
+    }
+
+    val askOldData = dbActor ? DbActor.ReadLaoData(rpcMessage.getParamsChannel)
+
+    rpcMessage.getParamsMessage match {
+      case Some(message: Message) =>
+        val data: CloseRollCall = message.decodedData.get.asInstanceOf[CloseRollCall]
+        val askCreateChannels: Future[GraphMessage] = createAttendeeChannels(data, rpcMessage)
+
+        // We create a new Future (tuple of Futures) to remove the obligatory sequential execution
+        // TODO: A possible addition would be to handle the error messages better
+        val resFuture: Future[GraphMessage] = (for{
+          f1 <- ask
+          f2 <- askOldData
+          f3 <- askCreateChannels
+        } yield(f1, f2, f3)).map{
+          case (Left(_), DbActorReadLaoDataAck(Some(_)), Left(_)) => Left(rpcMessage)
+          case (Left(_), DbActorNAck(code, description), Left(_)) => Right(PipelineError(code, description, rpcMessage.id))
+          case (Left(_), _, error@Right(_)) => error
+          case (error@Right(_), _, _) => error
+          case _ => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswer, rpcMessage.id))
+        }
+
+        val laoChannel: Option[Base64Data] = rpcMessage.getParamsChannel.decodeChannelLaoId
+        laoChannel match {
+          case None => Right(PipelineError(
+            ErrorCodes.SERVER_ERROR.id,
+            s"There is an issue with the data of the LAO",
+            rpcMessage.id
+          ))
+          case Some(_) => Await.result(resFuture, duration)
```

Comment:
doesn't this mean that `resFuture` will always be executed, even if the `rpcMessage` contains an invalid LAO channel ?

Reply:
Since the decodeChannelLaoId should return a None as soon as things go wrong, it should not be executed when the channel is invalid, but I can change the case order if it looks better that way (first Some(_) then None)

Reply:
You are both talking about something different I believe ^^

- @tuomas-p : changing the order of the cases will not change anything (except aesthetically/code comprehension) :)
- @pierluca (and @tuomas-p) : yes that is correct. At line 109-110, the `resFuture` will **already** be "in progress" but not necessarily finished. We distinguish two cases:
  - assuming we take the branch at line 117 (Some case), everything is gucci 😎 
  - assuming we take the branch at line 112 (None case), then:
    - if resFuture finishes (i.e. either Successes or Fails), my guess is that it will be "garbage collected" (i.e. no problem there)
    - if resFuture stays pending, then the future will keep running/calculating in the background until completion (which could take forever in some cases)

Reply:
PS: check my message that starts with "I would argue it's already good:" for a solution (screenshot)